store/sqlstore/upgrade.go

```diff
@@ -1208,6 +1209,201 @@ func upgradeDatabaseToVersion537(sqlStore *SqlStore) {
 	sqlStore.RemoveIndexIfExists("IDX_RetentionPolicies_DisplayName_Id", "RetentionPolicies")
 	sqlStore.CreateIndexIfNotExists("IDX_RetentionPolicies_DisplayName", "RetentionPolicies", "DisplayName")
 
+	fixCRTThreadCountsAndUnreads(sqlStore)
+	fixCRTChannelMembershipCounts(sqlStore)
+	fixCRTChannelUnreadsForJoinLeaveMessages(sqlStore)
+
 	// saveSchemaVersion(sqlStore, Version5370)
 	// }
 }
+
+// fixCRTThreadCountsAndUnreads
+// 1. Fixes reply counts, last reply times and participants of threads.
+// 2. Marks threads as read for users where the last reply time of the thread
+//    is earlier than the time the user viewed the channel.
+//    Marking a thread means setting the mention count to zero and
+//    setting the last viewed at time of the the thread as the last viewed at time
+//    of the channel
+func fixCRTThreadCountsAndUnreads(sqlStore *SqlStore) {
+	countsAndLastReplyCTE := `
+		SELECT RootId, COUNT(Id) AS FixedReplyCount, max(Posts.CreateAt) as FixedLastReply
+		FROM Posts where Posts.DeleteAt = 0
+		GROUP BY RootId
+	`
+	updateReplyCountsQuery := `
+		WITH q AS (` + countsAndLastReplyCTE + `)
+		UPDATE Threads
+		SET ReplyCount = q.FixedReplyCount, LastReplyAt = q.FixedLastReply
+		FROM q
+		WHERE Threads.PostId = q.RootId;
+	`
+
+	if sqlStore.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		updateReplyCountsQuery = `
+			UPDATE Threads
+			INNER JOIN (` + countsAndLastReplyCTE + `) as q
+			ON Threads.PostId = q.RootId
+			SET ReplyCount = q.FixedReplyCount, LastReplyAt = q.FixedLastReply
+		`
+	}
+
+	participantsCTE := `
+		SELECT COALESCE(NULLIF(RootId,''), Id) AS ThreadId, ARRAY_AGG(DISTINCT UserId) FixedParticipants
+		FROM Posts
+		WHERE Posts.DeleteAt = 0
+		GROUP BY ThreadId
```

Comment:
Same here, looks like we're grabbing participants for all threads instead of just the ones we care about?

Reply:
Yep, same. Need to batch.